climakitae/selectors.py

```diff
@@ -0,0 +1,222 @@
+    """
+
+    subset_by_lat_lon = param.Boolean()
+    # would be nice if these lat/lon sliders were greyed-out when subset option is not selected
+    latitude = param.Range(default=(41, 42), bounds=(10, 67))
+    longitude = param.Range(default=(-125, -115), bounds=(-156.82317, -84.18701))
+    # shapefile = param.FileSelector(path='../../*/*.shp*', precedence=0.5) #not for March 2022
+    # cached_area = param.ObjectSelector(objects=['CA','Sierra','LA County']) #this is a placeholder list
+
+    # @param.depends('cached_area',watch=True)
+    # def _update_loc_cached(self):
+    #    '''Updates the 'location' object to be the Polygon associated with the selected geographic area.'''
+    #    location = _areas_database[self.cached_area] #need this database to exist...
+
+    # not for soft launch:
+    # @param.depends('shapefile',watch=True)
+    # def _update_loc_shp(self):
+    #    '''Updates the 'location' object to be the polygon in the uploaded shapefile.
+    #    Dealing with user-uploaded data of any kind might not be in the soft launch.'''
+    #    # probably need to do checking for valid input (also for security reasons)
+    #    user_location = gpd.read_file("shapefile.shp")
+    #    assert user_location.geom_type in ['Point','Polygon'], "Please upload a valid shapefile."
+    #    # maybe also offer interactive selecting if there's more than one
+    #    # polygon in the shapefile!
+    #    location = user_location
+
+    # doesn't display yet for some reason:
+    @param.depends("latitude", "longitude", watch=True)
+    def view(self):
+        geometry = box(
+            self.longitude[0], self.latitude[0], self.longitude[1], self.latitude[1]
+        )
+        fig0 = plt.figure()
+        ax = plt.subplot(111, projection=ccrs.Orthographic(-115, 40))
+        ax.set_extent([-160, -84, 8, 68], crs=ccrs.PlateCarree())
+        ax.add_geometries([geometry], crs=ccrs.PlateCarree())
+        ax.coastlines()
+        ax.add_feature(cfeature.STATES, linewidth=0.5)
+        return pn.pane.Matplotlib(fig0, dpi=144)
+
+
+class LocSelectorPoint(param.Parameterized):
+    """
+    If the user wants a timeseries that pertains to a point, they may choose from among a set of
+    pre-calculated station locations. Later this class can be extended to accomodate user-specified
+    station data. Produces a panel from which to select from pre-calculated stations, and updates
+    the 'location' object accordingly.
+    """
+
+    cached_station = param.Selector(objects=_cached_stations)
+
+    @param.depends("cached_station", watch=True)
+    def _update_location(self):
+        """Updates the 'location' object to be the point associated with the selected station."""
+        location = _stations_database[self.cached_station]
+
+
+class CatalogContents:
+    def __init__(self):
+        # get the list of data variables from one of the zarr files:
+        cat = intake.open_catalog("s3://cdcat/cae.yaml")
```

Comment:
I get that we should try to load the catalog only once, but I'm not sure what you mean as a suggestion here "...so this could be set to an instance attribute here and referenced elsewhere which should cache the catalog and make additional calls faster."

Reply:
Are you talking about a nested class called Catalog inside Application?